### PR TITLE
add field Number (^FN) and recall / download format (^DF, ^XF)

### PR DESCRIPTION
**Pull request by MRejou**
I tried to match coding style and architecture of existing code.

-     Added FieldNumberCommandAnalyzer, DownloadFormatCommandAnalyzer, RecallFormatCommandAnalyzer
-     Added ZplFieldNumber, ZplRecallFieldNumber and ZplRecallFormat elements
-     Small refactor of ZplTextField do that other elements can print characters without inheritance (it may be converted to an external helper class, but I wanted to avoid heavy changes)
-     Updated FieldDataZplCommandAnalyzer to return an element matching ^FN...^FD... pair
-     Updated FieldSeparatorZplCommandAnalyzer to return an element matching trailing ^FN commands before ^FS
-     Added DownloadFormatName property in LabelInfo class
-     Added few properties to VirtualPrinter
-     Added format merging in IFormatMerger / FormatMerger, invoked from ZplAnalyzer
-     Bonus : added an interface IZplAnalyzer (allowing DI in apps based on this lib)

### CODE REVIEW

File: src/BinaryKits.Zpl.Viewer/ZplAnalyzer.cs [hunk 1]
```diff
@@ -5,19 +5,22 @@
...
         private readonly IPrinterStorage _printerStorage;
+        private readonly IFormatMerger _formatMerger;
         private readonly string _labelStartCommand = "^XA";
         private readonly string _labelEndCommand = "^XZ";
 
-        public ZplAnalyzer(IPrinterStorage printerStorage)
+        public ZplAnalyzer(IPrinterStorage printerStorage, IFormatMerger formatMerger = null)
         {
             this._printerStorage = printerStorage;
+            _formatMerger = formatMerger ?? new FormatMerger();
```
Comment (tinohager) [inline src/BinaryKits.Zpl.Viewer/ZplAnalyzer.cs:23]: add a this.

File: src/BinaryKits.Zpl.Viewer/ZplAnalyzer.cs [hunk 2]
```diff
@@ -63,16 +71,18 @@ public AnalyzeInfo Analyze(string zplData)
             {
                 var currentCommand = zplCommands[i];
```
Comment (tinohager) [inline src/BinaryKits.Zpl.Viewer/ZplAnalyzer.cs:72]: can we move the Trim to the top?

File: src/BinaryKits.Zpl.Viewer/FormatMerger.cs [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+{
+    /// <summary>
+    /// Helper class responsible of merging several LabelInfo based on format data (templating).
+    /// </summary>
+    public class FormatMerger : IFormatMerger
+    {
+        /// <inheritdoc />
+        public List<LabelInfo> MergeFormats(List<LabelInfo> rawLabelInfos)
+        {
+            List<LabelInfo> mergedLabelInfos = new List<LabelInfo>();
```
Comment (tinohager) [inline src/BinaryKits.Zpl.Viewer/FormatMerger.cs:18]: use var

File: src/BinaryKits.Zpl.Viewer/FormatMerger.cs [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+            // Format label infos indexed by download format name
+            Dictionary<string, LabelInfo> templateFormats = new Dictionary<string, LabelInfo>();
```
Comment (tinohager) [inline src/BinaryKits.Zpl.Viewer/FormatMerger.cs:20]: use var

File: src/BinaryKits.Zpl.Viewer/ZplAnalyzer.cs (hunk 2 above)
Comment (MRejou) [reply src/BinaryKits.Zpl.Viewer/ZplAnalyzer.cs:72]: hello !

I hesitated.

Indeed think we may share this, but I'm not 100% confident there is no risk for analyzer.Analyze(currentCommand) : 
I don't know how to ensure there will not be any side-effect :(

I propose a trimmedCommand only for start/end commands if you agree ?

**Pull request merged**